### PR TITLE
Add support for contributions through `.spec.contributions` field in DevWorkspace

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
Adds support for the `.spec.contributions` field as described in https://github.com/devfile/api/issues/874. The new field functions the same as current `plugin` components but can be defined separately from `.spec.template`, allowing easier conversion from devfile to devworkspace and back.

For test cases, I copied most of the existing tests for plugins functionality and adapted them to use contributions instead. 

### What issues does this PR fix or reference?
Part of https://github.com/devfile/devworkspace-operator/issues/656

### Is it tested? How?
This PR adds a new sample: `samples/theia-next-contributions.yaml` that uses the new field instead of a theia plugin component. This should behave identically to the current `theia-next.yaml` sample.

In addition I've rewritten the samples from #844 to use contributions instead of plugin components:
* checode + golang [[link]](https://gist.github.com/amisevsk/b44ab9d83c2ac6cb4c500ad16ee70ca7#file-code-golang-devfile-yaml):
    ```
    kubectl apply -f https://gist.githubusercontent.com/amisevsk/b44ab9d83c2ac6cb4c500ad16ee70ca7/raw/a4506801c3d20e6a1423a98238d7e8d9636005f3/code-golang.devfile.yaml
    ```

* theia + golang [[link]](https://gist.github.com/amisevsk/b44ab9d83c2ac6cb4c500ad16ee70ca7#file-theia-golang-devworkspace-yaml):
    ```
    kubectl apply -f https://gist.githubusercontent.com/amisevsk/b44ab9d83c2ac6cb4c500ad16ee70ca7/raw/a4506801c3d20e6a1423a98238d7e8d9636005f3/theia-golang.devworkspace.yaml
    ```

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

### CODE REVIEW

**Comment by AObuchow** [comment]
@amisevsk the code looks good to me, though in my testing on Minikube I found that the devworkspace URL does not appear with the `theia-next-contributions.yaml` sample or [checode + golang](https://gist.github.com/amisevsk/b44ab9d83c2ac6cb4c500ad16ee70ca7#file-code-golang-devfile-yaml).

```
[aobuchow@fedora samples]$ kubectl get dw -w
NAME                DEVWORKSPACE ID             PHASE      INFO
test-contrib-code   workspace30e25356b27043bc   Starting   Waiting for workspace deployment
test-contrib-code   workspace30e25356b27043bc   Starting   Waiting for workspace deployment
test-contrib-code   workspace30e25356b27043bc   Running    Workspace is running
```

Is there something I'm missing here?

**Comment by amisevsk** [comment]
@AObuchow Weird, that doesn't reproduce for me. Are you sure you're running the changes from this PR?

**Comment by AObuchow** [comment]
> @AObuchow Weird, that doesn't reproduce for me. Are you sure you're running the changes from this PR? New CRDs + old DWO image would explain that.

I deleted my minikube cluster and reinstalled everything, just to make sure I had a clean slate. 

I was able to get the `theia-next-contributions.yaml` sample working by building and deploying the DWO controller image to minikube (i.e. not scaling the deployment to 0) rather than running the controller locally. However, when running the controller locally, this issue still occurs. 

I checked the git log and diff'ed the CRD's and code changes and everything seems in place.

Are you testing with the controller running locally as well? If so, then it's something else on my end for sure.

**Comment by amisevsk** [comment]
@AObuchow doesn't reproduce when running locally either. Try a clean install: `make uninstall`, `make install`

**Comment by dkwon17** [comment]
I've only tested locally, and got the same behaviour as @AObuchow if I apply the template with `oc apply -f samples/theia-next-contributions.yaml`. For some reason, the `spec.contributions` field is removed when creating the DevWorkspace.

However, I'm able to do `oc edit devworkspace theia-next` and add the `spec.contributions` back manually, and that got it working; the workspace pod starts with the `theia-ide` container contributed from `spec.contributions`

**Comment by amisevsk** [comment]
@dkwon17 does this still occur if you run `make uninstall` before trying to test?

**Comment by amisevsk** [comment]
I've tested on OpenShift 4.11 and minikube (k8s 1.23.3) and could not reproduce this issue :thinking:

**Comment by dkwon17** [comment]
> In either case, make sure that the webhooks deployment for your test environment is running a PR branch build -- if your webhooks server is using the old DevWorkspace definitions, it's likely not deserializing the components field to begin with, causing it to be dropped when the mutating webhook adds a creator label.

Thanks @amisevsk , I realized this was the culprit. If I update the `webhook-server` image from the `devworkspace-webhook-server` deployment with the image for this PR, then I don't get the issue of having the `spec.contributions` disappearing

**Comment by amisevsk** [comment]
Thank goodness, I'm glad we figured it out -- I thought I was losing my mind! :smile:

**Review by dkwon17 — APPROVED** [review]
Just an optional suggestion, LGTM

**Review by ibuziuk — APPROVED** [review]

File: pkg/library/flatten/flatten.go [hunk 1]
```diff
@@ -75,49 +75,73 @@ func ResolveDevWorkspace(workspace *dw.DevWorkspaceTemplateSpec, tooling Resolve
...
-			pluginComponent, err := resolvePluginComponent(component.Name, component.Plugin, tooling)
+		resolvedPlugin, err := recursiveResolve(pluginComponent, nil, tooling, newCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		annotate.AddSourceAttributesForTemplate(contribution.Name, resolvedPlugin)
+		pluginSpecContents = append(pluginSpecContents, &resolvedPlugin.DevWorkspaceTemplateSpecContent)
+	}
+	if workspace != nil {
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/library/flatten/flatten.go:109]: Makes sense, thanks. Squashed change into earlier commit.

**Pull request merged**